### PR TITLE
Refactor pantie patch editor convert window

### PR DESCRIPTION
**Pull request by RamType0**
- Rx UI management
- Use UIElements in Unity2019.4 or newer

### CODE REVIEW

**Comment by RamType0** [comment]
Also refactored other editor windows.

File: Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs [hunk 1]
```diff
@@ -2,53 +2,66 @@
...
+        AsyncReactiveProperty<Texture> BaseAvatarTexture { get; } = new AsyncReactiveProperty<Texture>(null);
+        AsyncReactiveProperty<Texture> ConvertTexture { get; } = new AsyncReactiveProperty<Texture>(null);
+        AsyncReactiveProperty<int> SelectedAvatarIndex { get; } = new AsyncReactiveProperty<int>(-1);
+        AsyncReactiveProperty<bool> ConvertRunning { get; } = new AsyncReactiveProperty<bool>(false);
+
+        ReadOnlyAsyncReactiveProperty<bool> CanStartConvert { get; } 
         
+        public PantiePatchEditorConvertWindow()
+        {
+            CanStartConvert = UniTaskAsyncEnumerable.CombineLatest(ConvertTexture, SelectedAvatarIndex, ConvertRunning,
```
Comment (sansuke05) [inline Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs:31]: メモ：https://qiita.com/Marimoiro/items/9252cf7eaefee22a9f76

File: Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs [hunk 2]
```diff
@@ -2,53 +2,66 @@
...
 
-        private Gateway _gateway;
-        
-        private Texture convertTexture;
+#pragma warning disable CS0414//The field 'PantiePatchEditorConvertWindow.visualTree' is assigned but its value is never used
+        [SerializeField] VisualTreeAsset visualTree = default;
+#pragma warning restore
 
-        private Texture baseAvaterTexture;
+        AsyncReactiveProperty<Texture> BaseAvatarTexture { get; } = new AsyncReactiveProperty<Texture>(null);
```
Comment (sansuke05) [inline Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs:22]: メモ：https://qiita.com/toRisouP/items/8f66fd952eaffeaf3107#asyncreactiveproperty

File: Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs [hunk 3]
```diff
@@ -2,53 +2,66 @@
...
+        AsyncReactiveProperty<int> SelectedAvatarIndex { get; } = new AsyncReactiveProperty<int>(-1);
+        AsyncReactiveProperty<bool> ConvertRunning { get; } = new AsyncReactiveProperty<bool>(false);
+
+        ReadOnlyAsyncReactiveProperty<bool> CanStartConvert { get; } 
         
+        public PantiePatchEditorConvertWindow()
+        {
+            CanStartConvert = UniTaskAsyncEnumerable.CombineLatest(ConvertTexture, SelectedAvatarIndex, ConvertRunning,
+               (convertTexture, selectedAvatarIndex, convertRunning) => convertTexture != null && selectedAvatarIndex != -1 && !convertRunning).ToReadOnlyAsyncReactiveProperty(default);
+        }
```
Comment (sansuke05) [inline Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs:33]: メモ：https://cysharp.github.io/UniTask/api/Cysharp.Threading.Tasks.StateExtensions.html#Cysharp_Threading_Tasks_StateExtensions_ToReadOnlyAsyncReactiveProperty__1_Cysharp_Threading_Tasks_IUniTaskAsyncEnumerable___0____0_CancellationToken_

Comment (sansuke05) [reply Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs:33]: @RamType0 I have a question. Why are you doing the process of assigning to `CanStartConvert` in the constructor?

Comment (sansuke05) [reply Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs:33]: @RamType0 Is this the same as using `Observable.CombineLatest` ?

File: Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs [hunk 4]
```diff
@@ -78,37 +91,86 @@ public class PantiePatchEditorConvertWindow : EditorWindow {
...
+                        "アバターのデータがダウンロードされていません\nメニューのデータダウンロード > 対応アバター情報の更新からデータのダウンロードをして下さい",
+                        MessageType.Info
+                    );
+                }
+            };
+
+            var convertButton = root.Q<Button>();
+            convertButton.SetEnabled(false);
+
+            CanStartConvert.ForEachAsync((canStartConvert) => convertButton.SetEnabled(canStartConvert));
```
Comment (sansuke05) [inline Assets/AliceLaboratory/Editor/PantiePatchEditorConvertWindow.cs:146]: メモ：https://light11.hatenadiary.com/entry/2019/01/28/231639

File: Assets/AliceLaboratory/Editor/PantiePatchEditorDLWindow.cs [hunk 5]
```diff
@@ -52,42 +40,29 @@ public class PantiePatchEditorDLWindow : EditorWindow {
...
             // --- 変換元パンツテクスチャのリストをDL ---
-            var dreamsData = await _gateway.GetDreamsData();
+            var getDreamsData = Gateway.GetDreamsData();
+            Gateway.ShowProgressBarForUnityWebRequest(getDreamsData.Request, "Downloading...", "Downloading our dreams");
+            var dreamsData = await getDreamsData.Task;
             if (dreamsData == null)
             {
                 Debug.LogError("Download Error: 変換元パンツ情報のダウンロードに失敗しました");
-                ClearGUIWaiting();
+
```
Comment (sansuke05) [inline Assets/AliceLaboratory/Editor/PantiePatchEditorDLWindow.cs:65]: You'll exit this function on the next line, so you need `DownloadingDreams.Value = false;` here, right?

**Review by sansuke05 — COMMENTED** [review]
I commented. 😊

File: Assets/AliceLaboratory/Editor/PantiePatchEditorDLWindow.cs (hunk 5 above)
Comment (RamType0) [reply Assets/AliceLaboratory/Editor/PantiePatchEditorDLWindow.cs:65]: Exactly.

**Review by RamType0 — COMMENTED** [review]
Maybe I have a mistake at here.
We should do `DownloadingDreams.Value = false` with try-finally block.

File: Assets/AliceLaboratory/Editor/PantiePatchEditorDLWindow.cs (hunk 5 above)
Comment (RamType0) [reply Assets/AliceLaboratory/Editor/PantiePatchEditorDLWindow.cs:65]: Maybe I have a mistake at here.
We should do DownloadingDreams.Value = false with try-finally block.

**Comment by sansuke05** [comment]
@RamType0 マージ用のPRを作成したのでこちらはcloseします。

**Pull request closed without merging**